### PR TITLE
Ensure patch version is provided if accessing versioned documentation

### PR DESCRIPTION
**Pull request by nain-F49FF806**
This is step one in addressing https://github.com/rust-lang/simpleinfra/issues/360

Have modified the router slightly

- If complete (including patch) version provided, forward as is
- If only upto minor version is provided, redirect to patch zero of that minor version

This should help in redirecting minor version documentation access to (hopefully) valid pages

Todo/Future: Consider forwarding to latest patch when only minor version provided

### CODE REVIEW

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (nain-F49FF806) [inline terragrunt/modules/release-distribution/lambdas/doc-router/index.js:78]: Note: Temp redirect is used so the patch version used can be updated later.

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (MarcoIeni) [inline terragrunt/modules/release-distribution/lambdas/doc-router/index.js:70]: This regex looks good 👍
image

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (MarcoIeni) [inline terragrunt/modules/release-distribution/lambdas/doc-router/index.js:74]: I'm not sure if we can do this.
Asked in the original issue.
If we can't let's delete this line. 👍

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (MarcoIeni) [inline terragrunt/modules/release-distribution/lambdas/doc-router/index.js:75]: This regex looks good 👍
image

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (MarcoIeni) [reply terragrunt/modules/release-distribution/lambdas/doc-router/index.js:78]: why do we want to update it later? 🤔

**Comment by MarcoIeni** [comment]
Looks great 🥳
Let me check if this is all that's required and I'll deploy it to the dev environment

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (MarcoIeni) [inline terragrunt/modules/release-distribution/lambdas/doc-router/index.js:73]:
```suggestion
    // Include patch version 0 if minor version is provided
```

**Comment by MarcoIeni** [comment]
Applied to dev-doc.rust-lang.org 👍
However https://dev-doc.rust-lang.org/1.65/std/boxed/struct.Box.html it's not working

**Comment by MarcoIeni** [comment]
But https://dev-doc.rust-lang.org/1.65.0/std/boxed/struct.Box.html it's not working too.
let me check

**Comment by MarcoIeni** [comment]
I redeployed the master branch and apparently those URLs are not working in dev 🤔

**Comment by MarcoIeni** [comment]
Ok. Now https://dev-doc.rust-lang.org/1.65.0/std/boxed/struct.Box.html is working

**Comment by MarcoIeni** [comment]
I applied your code and https://dev-doc.rust-lang.org/1.65/std/boxed/struct.Box.html isn't working.

Do you have any idea? Maybe the regex substitution doesn't work?

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (nain-F49FF806) [reply terragrunt/modules/release-distribution/lambdas/doc-router/index.js:78]: This is in case we could update the code to get the latest patch live and direct to it. 
If we can then temp redirect would be good, as the latest patch version for a minor release can change. 

If we are happy with always redirecting to the patch zero release. Then regular redirect should do.

**Comment by nain-F49FF806** [comment]
> I applied your code and https://dev-doc.rust-lang.org/1.65/std/boxed/struct.Box.html isn't working.
> 
> Do you have any idea? Maybe the regex substitution doesn't work?


Hi, I checked the request in browser's dev console.
image


I suspect the pattern has not matched due to the leading `/` indicating the root. 
Have made a commit fixing the regex accordingly. 

Could you try now?


p.s: 
  If the above is the correct reason why it didn't work then we may have to update the earlier regex also. 
  Have edited only one regex now  so we can see if it works without changing behavior of the patch versioned case.

**Comment by MarcoIeni** [comment]
Deloyed 👍
https://dev-doc.rust-lang.org/1.65/std/boxed/struct.Box.html still doesn't work

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (MarcoIeni) [inline terragrunt/modules/release-distribution/lambdas/doc-router/index.js:74]:
```suggestion
    const minor_versioned_pattern = /^\/(\d+)\.(\d+)\/(.*)/
```
The first `/` is always there I think

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (MarcoIeni) [inline terragrunt/modules/release-distribution/lambdas/doc-router/index.js:70]:
```suggestion
    if (/^\/(\d+)\.(\d+).(\d+)\/(.*)/.test(request.uri)) {
```
There's a `/` at the beginning, right? It was there before.

Comment (nain-F49FF806) [reply terragrunt/modules/release-distribution/lambdas/doc-router/index.js]: Absolutely! I seem to have missed it. Have added it back now.

File: terragrunt/modules/release-distribution/lambdas/doc-router/index.js
Comment (nain-F49FF806) [reply terragrunt/modules/release-distribution/lambdas/doc-router/index.js]: You are correct! It is always [required](https://www.rfc-editor.org/rfc/rfc7230#section-5.3.1) to be present.
Have edited accordingly.

**Comment by jdno** [comment]
This is something that I'd love to see added to our [rust-lang/infra-smoke-tests](https://github.com/rust-lang/infra-smoke-tests). It will probably require creating a new `TestSuite`, so I'll try to create a tracking issue with some instructions later.

**Comment by MarcoIeni** [comment]
Looks great, I'm testing this one last time just to be sure 😂

**Review by MarcoIeni — APPROVED** [review]
This is working for me and the code looks good. 👍
Great job, thanks 🙌
I'll wait for a review from JD since CDN is quite critical.

**Comment by MarcoIeni** [comment]
One note for the maintainers: I want to merge this with "squash" strategy. Is that fine or do we want to the contributors to force push?

**Review by jdno — APPROVED** [review]
Looks good! I'm adding it to our [smoke tests](https://github.com/rust-lang/infra-smoke-tests), but we can already go ahead and deploy it...

**Comment by MarcoIeni** [comment]
nain, can you squash your commits into a single one? After that, push a new commit where you update [this](https://github.com/rust-lang/simpleinfra/blob/master/terragrunt/accounts/legacy/releases-prod/deployed-ref#L1) sha with the sha of your first commit.

So in total, this PR should contain 2 commits.

**Comment by MarcoIeni** [comment]
Applied to production 👍

**Comment by MarcoIeni** [comment]
https://doc.rust-lang.org/1.65/std/boxed/struct.Box.html it's working 🎉

Thanks nain and sorry if it took us so long to review but CDNs are quite critical!

I hope to see you again in my GitHub notifications 😁

**Pull request merged**
